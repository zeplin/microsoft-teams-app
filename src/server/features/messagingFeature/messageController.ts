import { RequestHandler } from "express";
import { messageFacade } from "./messageFacade";

export const handleWebhookRequest: RequestHandler = async (req, res, next) => {
    try {
        const webhookId = req.headers["zeplin-webhook-id"] as string;
        const deliveryId = req.headers["zeplin-delivery-id"] as string;
        await messageFacade.handleEventArrived({
            webhookId,
            deliveryId,
            payload: req.body
        });
        res.send();
    } catch (error) {
        next(error);
    }
};
