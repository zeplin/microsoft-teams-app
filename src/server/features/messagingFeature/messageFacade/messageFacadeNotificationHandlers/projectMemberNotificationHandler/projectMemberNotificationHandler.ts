import { NotificationHandler } from "../NotificationHandler";
import { commonTeamsCard, MessageCard } from "../teamsCardTemplates";
import { ProjectMemberInviteEvent, WebhookEvent } from "../../../../../adapters/zeplin/types";
import { LONG_DELAY } from "../constants";

class ProjectMemberNotificationHandler extends NotificationHandler<ProjectMemberInviteEvent> {
    delay = LONG_DELAY;

    private getText(events: ProjectMemberInviteEvent[]): string {
        const [{
            payload: {
                context: {
                    project: {
                        name: projectName
                    }
                },
                resource: {
                    data: {
                        user: {
                            username
                        }
                    }
                }
            }
        }] = events;
        return events.length === 1
            ? `**${username}** just joined _${projectName}_.`
            : `**${events.length} new users** just joined _${projectName}_`;
    }

    getTeamsMessage(events: ProjectMemberInviteEvent[]): MessageCard {
        return commonTeamsCard({
            text: this.getText(events),
            section: {
                text: "Say hi ðŸ‘‹"
            }
        });
    }

    shouldHandleEvent(event: WebhookEvent): event is ProjectMemberInviteEvent {
        return event.payload.action === "invited";
    }
}

export const projectMemberNotificationHandler = new ProjectMemberNotificationHandler();
