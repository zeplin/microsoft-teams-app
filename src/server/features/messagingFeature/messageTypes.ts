type EventDescriptor = {
    type: string;
    action: string;
}

type Resource = {
    id: string;
    type: ResourceType;
    data?: object;
}

export interface ProjectContext {
    project: {
        id: string;
    };
}

export enum ResourceType {
    PROJECT = "Project",
    STYLEGUIDE = "Styleguide",
    COLOR = "Color",
    TEXT_STYLE = "TextStyle",
    COMPONENT = "Component",
    SPACING_TOKEN = "SpacingToken",
    PROJECT_MEMBER = "ProjectMember",
    STYLEGUIDE_MEMBER = "StyleguideMember",
    ORGANIZATION_SUMMARY = "OrganizationSummary",
    SCREEN = "Screen",
    SCREEN_VERSION = "ScreenVersion",
    ORGANIZATION_MEMBER = "OrganizationMember",
    SCREEN_NOTE = "ScreenNote",
    SCREEN_NOTE_COMMENT = "ScreenNoteComment"
}

export interface EventPayload<
    E extends EventDescriptor,
    C extends object | undefined,
    R extends Resource,
> {
    event: E["type"];
    action: E["action"];
    timestamp: number;
    context: C;
    resource: R;
    actor: {
        user: {
            id: string;
            email: string;
            username?: string;
            emotar?: string;
            avatar?: string;
        };
    };
}
export type CommonEventPayload = EventPayload<EventDescriptor, object | undefined, Resource>;
export type WebhookEvent<T extends CommonEventPayload = CommonEventPayload> = {
    webhookId: string;
    deliveryId: string;
    payload: T;
};
export type MessageJobData = {
    id: string;
    groupingKey: string;
}