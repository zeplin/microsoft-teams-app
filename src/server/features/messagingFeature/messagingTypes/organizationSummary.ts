export interface OrganizationSummary {
    id: string;
    name: string;
    logo?: string;
}