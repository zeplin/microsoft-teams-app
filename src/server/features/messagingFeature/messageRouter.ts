import { Router as createRouter } from "express";
import { handleWebhookRequest } from "./messageController";
import { JSONBodyParser } from "../../middlewares";

const messageRouter = createRouter({ mergeParams: true });
messageRouter.post("/",
    JSONBodyParser,
    handleWebhookRequest
);

export {
    messageRouter
};
