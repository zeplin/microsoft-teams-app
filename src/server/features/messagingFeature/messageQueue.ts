import Bull from "bull";
import { MessageJobData } from "./messageTypes";
import { Config } from "../../../config";

const QUEUE_NAME = "process-events";

type AddOptions = {
    delay: number;
}

class MessageQueue {
    private bullQueue: Bull.Queue<MessageJobData> | undefined;
    init(config: Config): void {
        this.bullQueue = new Bull<MessageJobData>(QUEUE_NAME, config.REDIS_URL);
    }

    process(handler: Bull.ProcessCallbackFunction<MessageJobData>): Promise<void> {
        return this.bullQueue.process(handler);
    }

    async add(data: MessageJobData, { delay }: AddOptions): Promise<void> {
        await this.bullQueue.add(data, {
            delay
        });
    }
}

export const messageQueue = new MessageQueue();
