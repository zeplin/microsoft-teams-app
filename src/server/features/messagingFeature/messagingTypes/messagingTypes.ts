import { OrganizationSummary } from "./organizationSummary";
import { RemPreferences } from "./remPreferences";

type EventDescriptor = {
    type: EventType;
    action: string;
}

type Resource = {
    id: string;
    type: ResourceType;
    data?: object;
}

export enum EventType {
    PROJECT_COLOR = "project.color",
    STYLEGUIDE_COLOR = "styleguide.color",
    PROJECT_NOTE = "project.note"
}

export interface ProjectContext {
    project: {
        id: string;
        created: number;
        name: string;
        description?: string;
        thumbnail?: string;
        platform: "android" | "ios" | "web" | "macos";
        status: "active" | "archived" | "deleted";
        organization?: OrganizationSummary;
        scene_url?: string;
        updated?: number;
        rem_preferences?: RemPreferences;
        number_of_screens: number;
        number_of_components: number;
        number_of_text_styles: number;
        number_of_colors: number;
        number_of_spacing_tokens: number;
        number_of_members: number;
        linked_styleguide?: {
            id: string;
        };
    };
}

export interface ScreenContext {
    screen: {
        id: string;
        created: number;
        updated?: number;
        tags: string[];
        name: string;
        image: {
            width: number;
            height: number;
            original_url?: string;
        };
        section?: {
            id: string;
        };
        description?: string;
        number_of_versions: number;
        number_of_notes: number;
    };
}

export interface StyleguideContext {
    styleguide: {
        id: string;
        created: number;
        name: string;
        description?: string;
        thumbnail?: string;
        platform: "base" | "web" | "ios" | "android" | "macos";
        status: "active" | "archived" | "deleted";
        organization?: OrganizationSummary;
        updated?: number;
        rem_preferences?: RemPreferences;
        number_of_components: number;
        number_of_text_styles: number;
        number_of_colors: number;
        number_of_spacing_tokens: number;
        number_of_members: number;
        parent?: {
            id: string;
        };
    };
}

export enum ResourceType {
    PROJECT = "Project",
    STYLEGUIDE = "Styleguide",
    COLOR = "Color",
    TEXT_STYLE = "TextStyle",
    COMPONENT = "Component",
    SPACING_TOKEN = "SpacingToken",
    PROJECT_MEMBER = "ProjectMember",
    STYLEGUIDE_MEMBER = "StyleguideMember",
    ORGANIZATION_SUMMARY = "OrganizationSummary",
    SCREEN = "Screen",
    SCREEN_VERSION = "ScreenVersion",
    ORGANIZATION_MEMBER = "OrganizationMember",
    SCREEN_NOTE = "ScreenNote",
    SCREEN_NOTE_COMMENT = "ScreenNoteComment"
}

export type User = {
    id: string;
    email: string;
    username: string;
    emotar?: string;
    avatar?: string;
}

export interface EventPayload<
    E extends EventDescriptor,
    C extends ProjectContext | StyleguideContext,
    R extends Resource,
> {
    event: E["type"];
    action: E["action"];
    timestamp: number;
    context: C;
    resource: R;
    actor: {
        user: User;
    };
}

export type CommonEventPayload = EventPayload<EventDescriptor, ProjectContext | StyleguideContext, Resource>;

export type WebhookEvent<T extends CommonEventPayload = CommonEventPayload> = {
    webhookId: string;
    deliveryId: string;
    payload: T;
};

export type MessageJobData = {
    id: string;
    groupingKey: string;
}