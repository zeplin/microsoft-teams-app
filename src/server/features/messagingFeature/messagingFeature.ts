import { messageQueue } from "./messageQueue";
import { messageFacade } from "./messageFacade";
import { Router } from "express";
import { messageRouter } from "./messageRouter";
import { Config } from "../../config";

export function initMessagingFeature(router: Router, config: Config): void {
    messageQueue.init(config);
    messageQueue.process(async job => {
        try {
            await messageFacade.processJob(job.data);
        } catch (err) {
            // TODO: Send to sentry
            // eslint-disable-next-line no-console
            console.error(err);
        }
    });
    router.use("/webhook", messageRouter);
}
