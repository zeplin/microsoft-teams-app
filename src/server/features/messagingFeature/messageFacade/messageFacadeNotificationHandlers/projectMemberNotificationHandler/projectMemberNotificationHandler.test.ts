import { WebhookEvent } from "server/features/messagingFeature/messagingTypes";
import { ProjectMemberEventPayload, projectMemberNotificationHandler } from "./projectMemberNotificationHandler";

type GetDummyEventParams = {
    projectName?: string;
    userId?: string;
    userName?: string;
}

function getDummyEvent({
    projectName = "Project MS Teams Integration",
    userId = "userId",
    userName = "dirtybit"
}: GetDummyEventParams = {}): WebhookEvent<ProjectMemberEventPayload> {
    return {
        payload: {
            action: "invited",
            context: {
                project: {
                    id: "projectId",
                    name: projectName
                }
            },
            resource: {
                data: {
                    user: {
                        id: userId,
                        username: userName
                    }
                }
            }
        }
    } as WebhookEvent<ProjectMemberEventPayload>;
}

describe("projectMemberNotificationHandler", () => {
    describe("getTeamsMessage", () => {
        it("should match snapshot when there is only one event", () => {
            expect(projectMemberNotificationHandler.getTeamsMessage([getDummyEvent()])).toMatchSnapshot();
        });

        it("should match snapshot when there are more than one events", () => {
            expect(projectMemberNotificationHandler.getTeamsMessage([
                getDummyEvent(),
                getDummyEvent({
                    userId: "userId2",
                    userName: "ergunsh"
                })
            ])).toMatchSnapshot();
        });
    });
});