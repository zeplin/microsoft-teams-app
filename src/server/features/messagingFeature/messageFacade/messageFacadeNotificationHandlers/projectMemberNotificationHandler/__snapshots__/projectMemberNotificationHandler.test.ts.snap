// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectMemberNotificationHandler getTeamsMessage should match snapshot when there are more than one events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 new users** just joined _Project MS Teams Integration_",
  "sections": [
    {
      "text": "Say hi ðŸ‘‹",
      "images": []
    }
  ]
}
`;

exports[`projectMemberNotificationHandler getTeamsMessage should match snapshot when there is only one event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**dirtybit** just joined _Project MS Teams Integration_.",
  "sections": [
    {
      "text": "Say hi ðŸ‘‹",
      "images": []
    }
  ]
}
`;
